### PR TITLE
[Editorial] Rewrite the section on member pointers

### PR DESCRIPTION
**Pull request by rjmccall**
Builds on #96.

- Break it up into subsections for data and function member pointers
- Be much more precise
- Use portable concepts
- Include the ARM representation for virtual member functions pointers
- Add a future-platform recommendation about null data member pointers

### CODE REVIEW

**Comment by rjmccall** [comment]
There's some prior of the change here on #95.

**Comment by rjmccall** [comment]
@zygoloid Any further comments on the current draft?

File: abi.html [hunk 1]
```diff
@@ -662,40 +693,110 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
...
-<dd> For a non-virtual function, this field is a simple function pointer.
-     (Under current base Itanium psABI conventions,
-     that is a pointer to a GP/function address pair.)
-     For a virtual function,
-     it is 1 plus the virtual table offset (in bytes) of the function,
-     represented as a <code>ptrdiff_t</code>.
-     The value zero represents a NULL pointer,
-     independent of the adjustment field value below.
+A null data member pointer is represented as an offset of <code>-1</code>.
+Unfortunately, it is possible to generate an data member pointer with an
```
Comment (zygoloid) [inline abi.html:710]: an -> a

File: abi.html [hunk 2]
```diff
@@ -662,40 +693,110 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
...
+offset of <code>-1</code> using explicit derived-to-base conversions.
+If this is done, implementations following this ABI may misbehave.
+<span class="future-abi">Recommendation for new platforms: consider using
+<code>PTRDIFF_MIN</code> as the null data member pointer instead.</span>
```
Comment (zygoloid) [inline abi.html:714]: I'm not sure whether we should be recommending this (and nothing else). Pointer-to-data-member casts generally need a branch or conditional move on nullness; is `PTRDIFF_MIN` easy enough to detect on most targets? (On x86, I suppose `cmp a 1; cmovno adjusted_value a` works.) If we're suggesting changing the representation anyway, I also think we should suggest a change that makes zero-initialization cheaper.

In particular, I think I would want people bringing up new targets to also consider storing the value left-shifted one place, using lowest bit == 0 for null pointers and lowest bit == 1 for non-null pointers. Pros: zero-initialization is all zeroes. Casts can be performed by unconditional adds. Null checks are fast. Con: require a (signed) right shift on every use.

File: abi.html [hunk 3]
```diff
@@ -662,40 +693,110 @@ <h3><a href="#pod"> 2.2 POD Data Types </a></h3>
...
+to never have its lowest bit set.  On most platforms, this is either
+always true, or it can be made true at little cost.  (For example,
+on most platforms a function pointer is just the address of the
+first instruction in the function.  Many architectures align all
+instructions to at least 2 bytes; even on architectures where that's
+not true, or where functions aren't byte-addressed, the compiler can
+just increase the alignment of non-static member functions until the
+low bit of the address is always reliably zero.)  However, some platforms
+use the low bit of a function pointer for special purposes, such as to
+distinguish THUMB functions on ARM.  Such platforms must use a slightly
```
Comment (zygoloid) [inline abi.html:778]: "must" seems a bit strong. How about "For such platforms, we recommend using a slightly different representation" or similar?

Comment (rjmccall) [reply abi.html]: I ended up rewriting this whole section.

File: abi.html (hunk 2 above)
Comment (rjmccall) [reply abi.html]: Yeah, good point.  Made a different recommendation.

**Comment by rjmccall** [comment]
I'll wait for feedback on the current draft.  My eventual goal (besides just improving the ABI document) is to build on this to document what arm64e does, which is to use dispatch thunks for virtual member function pointers.  However, there's no timeline attached to getting that in.

**Pull request merged**